rfcs/0134-nix-store-layer.md

```diff
@@ -0,0 +1,271 @@
+The division between `libnixstore`, `libfetchers`, `libexpr`, etc. is not yet exposed to users, emphasized in documentation (though this is changing thanks to @Fricklerhandwerk's efforts with new [architecture documentation](https://github.com/NixOS/nix/pull/6888)!).
+
+We should instead fully embrace it:
+
+- Docs
+
+  - More advanced documentation can explain layering for those that want a deeper understanding of Nix.
+
+  - Even more basic documentation can still benefit from separate terminology before the layering is fully explained.
+    See https://www.haskellforall.com/2022/08/stop-calling-everything-nix.html for a phenomenal take-down of how calling everything "Nix" today confuses users and leaves them unable to articulate what parts of the ecosystem are frustrating then.
+
+- Separate executable ensure lower layers build in isolation, new integration tests those executable without aid of high-layer info.
+
+### Starting with the store layer
+
+Ultimately, I would like to take this approach to all the layers.
+But I want to constrain the scope of this RFC to keep it tight an actionable.
+
+Layering between e.g. Flakes and the Nix Language doesn't yet exist in the implementation in the form of a library separation.
+I don't want to be "blocked" on major new development work, and anything involving Flakes is also far more controversial, and best avoided at first.
+
+Conversely, the store layer is already quite well separated.
+The orthogonality between it and other layers "proven" in the wild by projects like Guix (more on that latter part in the next section).
+The gap is also widest in terms of the layer of abstraction between, on one hand, nascent "infrastructure projects" like
+
+- CA derivations
+- Secrets in the store
+- Trustless remote building
+- Daemon and Hydra protocol rationalization
+- Windows Support
+- Computed derivations (which are built by the other derivations)
+- IPFS store
+
+etc. and "UX projects" like
+
+- Flakes
+- TOML Flakes
+- Hard-wired module system
+
+So focusing on the lowest layer first, we get the most "bang for buck" in terms of managing extremely different sorts of work separately.
+
+## Marketplace of Ideas
+
+As Nix grows more popularity, it will be inevitable that different groups want to explore in different directions.
+This is the *pluralism* of a larger community, and we should embrace it as a strength.
+We do that be fostering a *marketplace of ideas*.
+
+There are many possible ways in which to write down packages, The Nix language and Nixpkgs idioms, and Guix, for example, are just two points in a much larger space.
+There are also many possible ways set up build farms.
+Our current central dispatcher, many remote-builder agents model, point-to-point protocol model is also just point in a much larger design space.
+
+The "derivation language" and store *interface* however, seems to me at least to be a very natural design.
```

Comment:
By derivation language, you mean the ATerm-based language that drv files are written in?

Reply:
The abstract syntax of that. Basically "a derivation is an execve system call arguments, input black box data, and input other drvs' outputs". I think that right there is a very natural design.